### PR TITLE
Scope API keys to a gem (Take 3)

### PR DESCRIPTION
**Pull request by jenshenny**
Still may require a bit of polish but it's mostly all there. Will comment on parts of the code to clarify confusing bits

### Background
This PR allows users to select a gem scope an API key when creating and editing. Keys with this scope can only push, yank, add/remove owners with only the selected gem. All other behaviour would remain the same.

### Data model additions
New join table `ApiKeyRubygemScope` that belong to a gem ownership and an API key. An API key can up to one `ApiKeyRubygemScope`, ownerships can have many `ApiKeyRubygemScope`.

Scrappy drawing below
Screen Shot 2022-02-22 at 2 32 18 PM

There's also a `soft_deleted` column on ApiKeys, which will be explained more in the `What happens when an ownership is deleted?` section.

### Creating/Editing a Key with a Gem scope
If user finds themselves submitting an invalid rubygem id, an error will be displayed "Selected gem cannot be scoped to this key"

https://user-images.githubusercontent.com/42748004/161677305-ccff2bb3-b211-4627-a1bd-7e238e3e2133.mov

### CLI

https://user-images.githubusercontent.com/42748004/161677288-50109468-828a-4127-a272-4d94924a8187.mov

### What happens when an ownership is deleted?
When an ownership is deleted, it dependently destroys all associated ApiKeyRubygemScopes. In turn, in a before_destroy callback, the ApiKeyRubygemScope updates the `soft_deleted_at` associated ApiKey. Soft deleted API keys cannot be edited or used. The user can only delete it.

Screen Shot 2022-04-05 at 12 12 16 AM

https://user-images.githubusercontent.com/42748004/161677262-2e61ce73-b188-487b-8776-d5b3276d4f5b.mov

### What Reviewers should focus on
- `ApiKeyRubygemScope` name - is there a better name?
- `soft_deleted_at` - is there a better name? Thought about invalid but that can be confusing will Rails' validations
- Better way to set the associated rubygem on the API key? I went with raising and rescuing a RecordNotFound error. Thought about adding an error onto the model, but calling validations again (eg. save, valid?) removes the error.
- Better name for CLI error "Rubygem cannot be scoped to this API key"

### CODE REVIEW

File: app/controllers/api_keys_controller.rb [hunk 1]
```diff
@@ -15,7 +15,8 @@ def new
 
   def create
     key = generate_unique_rubygems_key
-    @api_key = current_user.api_keys.build(api_key_params.merge(hashed_key: hashed_key(key)))
+    build_params = { user: current_user, hashed_key: hashed_key(key), **api_key_params }
+    @api_key = ApiKey.new(build_params)
```
Comment (jenshenny) [inline app/controllers/api_keys_controller.rb:19]: setting `rubygem_id` is dependant on having user available, calling `build` on current_user doesn't set the user until the rest of the params have been set.

File: app/models/api_key.rb [hunk 2]
```diff
@@ -42,4 +59,8 @@ def other_enabled_scopes?
   def scope_presence
     errors.add :base, "Please enable at least one scope" unless enabled_scopes.any?
   end
+
+  def not_soft_deleted?
+    errors.add :base, "Key is invalid. Please delete and create a new one." if soft_deleted?
+  end
```
Comment (jenshenny) [inline app/models/api_key.rb:65]: Added a validation here but practically it should never be called. May be redundant?

File: app/models/api_key_rubygem_scope.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class ApiKeyRubygemScope < ApplicationRecord
+  belongs_to :api_key
+  belongs_to :ownership
+  validates :ownership_id, uniqueness: { scope: :api_key_id }
+  validates :api_key, :ownership, presence: true
+  before_destroy :soft_delete_api_key!, if: :destroyed_by_association
```
Comment (jenshenny) [inline app/models/api_key_rubygem_scope.rb:6]: Without the `if: destroyed_by_association`, changing a gem scope from a gem to all gems calls `soft_delete_api_key!`

File: db/schema.rb [hunk 4]
```diff
@@ -10,8 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2021_11_30_050124) do
-
+ActiveRecord::Schema[7.0].define(version: 2022_03_29_203956) do
```
Comment (tomstuart) [inline db/schema.rb:13]: Why do we need to update the schema to Rails 7? Is this required for this PR?

File: app/models/api_key.rb (hunk 2 above)
Comment (jchestershopify) [reply app/models/api_key.rb:65]: Some of the best validations are the ones you almost never need.

File: app/controllers/api/base_controller.rb [hunk 5]
```diff
@@ -25,6 +25,12 @@ def enqueue_web_hook_jobs(version)
     end
   end
 
+  def verify_api_key_gem_scope
+    return unless @api_key.rubygem && @api_key.rubygem != @rubygem
+
+    render plain: "Rubygem cannot be scoped to this API key", status: :forbidden
```
Comment (jchestershopify) [inline app/controllers/api/base_controller.rb:31]: I'd suggest `Gem` or `RubyGem` instead of `Rubygem`.

File: app/controllers/api/base_controller.rb [hunk 6]
```diff
@@ -55,4 +62,8 @@ def render_api_key_forbidden
       format.yaml { render yaml: { error: t(:api_key_forbidden) }, status: :forbidden }
     end
   end
+
+  def render_soft_deleted_api_key
+    render plain: "An invalid API key cannot be used. Please delete it and create a new one.", status: :forbidden
```
Comment (jchestershopify) [inline app/controllers/api/base_controller.rb:67]: Are we able to say why it's invalid?

File: app/models/api_key.rb [hunk 7]
```diff
@@ -42,4 +59,8 @@ def other_enabled_scopes?
   def scope_presence
     errors.add :base, "Please enable at least one scope" unless enabled_scopes.any?
   end
+
+  def not_soft_deleted?
+    errors.add :base, "Key is invalid. Please delete and create a new one." if soft_deleted?
```
Comment (jchestershopify) [inline app/models/api_key.rb:64]: This is different from the error message in `BaseController`, is that intentional?

File: app/models/pusher.rb [hunk 8]
```diff
@@ -3,22 +3,29 @@
...
   end
 
   def authorize
     rubygem.pushable? || rubygem.owned_by?(user) || notify_unauthorized
   end
 
+  def verify_gem_scope
+    return true unless @scoped_rubygem && rubygem != @scoped_rubygem
+
+    notify("Rubygem cannot be scoped to this API key", 403)
```
Comment (jchestershopify) [inline app/models/pusher.rb:26]: I think this should say `Gem` or `RubyGem`.

File: db/schema.rb [hunk 9]
```diff
@@ -27,10 +34,11 @@
     t.boolean "remove_owner", default: false, null: false
     t.boolean "access_webhooks", default: false, null: false
     t.boolean "show_dashboard", default: false, null: false
-    t.datetime "last_accessed_at"
-    t.datetime "created_at", null: false
-    t.datetime "updated_at", null: false
+    t.datetime "last_accessed_at", precision: nil
+    t.datetime "created_at", precision: nil, null: false
+    t.datetime "updated_at", precision: nil, null: false
```
Comment (jchestershopify) [inline db/schema.rb:39]: I think this is from running with Rails 7 tooling, I would leave it out.

Comment (jchestershopify) [reply db/schema.rb:39]: Same for other cases of `precision: nil` in the schema.

File: db/schema.rb [hunk 10]
```diff
@@ -84,20 +92,20 @@
...
-    t.string "code"
-    t.string "bugs"
-    t.datetime "created_at"
-    t.datetime "updated_at"
+    t.string "home", limit: 255
+    t.string "wiki", limit: 255
+    t.string "docs", limit: 255
+    t.string "mail", limit: 255
+    t.string "code", limit: 255
+    t.string "bugs", limit: 255
```
Comment (jchestershopify) [inline db/schema.rb:106]: Rails 7 again maybe? I don't think we want these limits, URLs can technically be arbitrarily long.

File: db/schema.rb [hunk 11]
```diff
@@ -117,8 +125,8 @@
     t.bigint "user_id"
     t.text "note"
     t.boolean "status", default: true, null: false
-    t.datetime "created_at", precision: 6, null: false
-    t.datetime "updated_at", precision: 6, null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
```
Comment (jchestershopify) [inline db/schema.rb:129]: I think we want to preserve the original precision.

File: db/schema.rb [hunk 12]
```diff
@@ -222,31 +230,31 @@
...
-    t.string "full_name"
+    t.string "full_name", limit: 255
+    t.string "licenses", limit: 255
     t.integer "size"
-    t.string "licenses"
     t.text "requirements"
-    t.string "required_ruby_version"
-    t.string "sha256"
+    t.string "required_ruby_version", limit: 255
+    t.string "sha256", limit: 255
```
Comment (jchestershopify) [inline db/schema.rb:249]: I don't think a limit adds a lot of value here, given the way PostgreSQL already intelligently handles strings.

File: test/functional/api/v1/rubygems_controller_test.rb [hunk 13]
```diff
@@ -528,6 +528,51 @@ def self.should_respond_to(format)
...
+        post :create, body: gem_file("test-1.0.0.gem").read
+      end
+
+      should respond_with :forbidden
+      should "#render_soft_deleted_api_key and display an error" do
+        assert_equal "An invalid API key cannot be used. Please delete it and create a new one.", @response.body
+      end
+    end
+
+    context "to not the gem being pushed" do
```
Comment (jchestershopify) [inline test/functional/api/v1/rubygems_controller_test.rb:548]: I found this a little hard to parse. Is there another phrasing?

**Review by jchestershopify — APPROVED** [review]
I noticed what looks like some Rails 7-related changes and bikeshedded a little on wording. But otherwise looks like a solid approach to me.

File: app/models/pusher.rb (hunk 8 above)
Comment (jenshenny) [reply app/models/pusher.rb:26]: Yeah the wording on this message could be improved. How does "This API key cannot perform this action on this gem." sound?

File: app/models/api_key.rb (hunk 7 above)
Comment (jenshenny) [reply app/models/api_key.rb:64]: It was intentional as I thought the message in `BaseController` wouldn't make sense here. On second thought however, it does fit - I'll make both messages the same 👍

File: app/controllers/api/base_controller.rb (hunk 6 above)
Comment (jenshenny) [reply app/controllers/api/base_controller.rb:67]: We could by process of elimination. Since gem ownership deletion is the only way to invalidate an api key, we know it's the reason. I don't think it's necessary to give more information, wdyt?

File: db/schema.rb (hunk 4 above)
Comment (jenshenny) [reply db/schema.rb:13]: To some degree yes. I recreated the migrations on the current master (which uses Rails 7). This created Rails 7.0 migrations which made these changes to the schema when db:migrate was ran. I do intend to open this as a separate PR in rubygems.org as it would be needed at some point.

**Review by tomstuart — APPROVED** [review]
Thanks Jenny! Let’s upstream! :shipit:
